src/dag/write.rs

```diff
@@ -1,55 +1,195 @@
-use super::chunk::Chunk;
 use super::key::Key;
+use super::{chunk::Chunk, meta_generated::meta};
 use super::{read, Result};
 use crate::kv;
+use async_recursion::async_recursion;
+use async_std::sync::RwLock;
+use futures::future::try_join_all;
+use futures::try_join;
+use std::collections::HashSet;
+
+#[derive(Debug, Default)]
+struct HeadChange(Option<String>, Option<String>);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum RefCount {
+    Inc = 1,
+    Dec = -1,
+}
 
 pub struct Write<'a> {
     kvw: Box<dyn kv::Write + 'a>,
+    changed_heads: RwLock<Vec<HeadChange>>,
+    mutated_chunks: RwLock<HashSet<String>>,
 }
 
 impl<'a> Write<'_> {
     pub fn new(kvw: Box<dyn kv::Write + 'a>) -> Write {
-        Write { kvw }
+        Write {
+            kvw,
+            changed_heads: Default::default(),
+            mutated_chunks: Default::default(),
+        }
     }
 
     pub fn read(&self) -> read::Read {
         read::Read::new(self.kvw.as_read())
     }
 
     pub async fn put_chunk(&mut self, c: &Chunk) -> Result<()> {
+        // TODO: These can be done in parallel.
+
         self.kvw
             .put(&Key::ChunkData(c.hash()).to_string(), c.data())
             .await?;
+
         if let Some(meta) = c.meta() {
             self.kvw
                 .put(&Key::ChunkMeta(c.hash()).to_string(), meta)
                 .await?;
         }
+
+        self.mutated_chunks
+            .write()
+            .await
+            .insert(c.hash().to_string());
+
         Ok(())
     }
 
     pub async fn set_head(&mut self, name: &str, hash: Option<&str>) -> Result<()> {
+        let old_hash = self.read().get_head(name).await?;
+
+        // TODO: These can be done in parallel.
+
+        let head_key = Key::Head(name).to_string();
         match hash {
-            None => Ok(self.kvw.del(&Key::Head(name).to_string()).await?),
-            Some(h) => Ok(self
-                .kvw
-                .put(&Key::Head(name).to_string(), h.as_bytes())
-                .await?),
+            None => self.kvw.del(&head_key).await?,
+            Some(h) => self.kvw.put(&head_key, h.as_bytes()).await?,
         }
+
+        self.changed_heads
```

Comment:
What happens if the same head gets set twice in the same tx? Seems like you want a map by head name.

Reply:
Nope, we need to keep track of that. Inc, Dec, Inc for example. In practice I don't think we will be setting head more than once per commit but this way we are sure that we do not end up clobbering old head changes

Reply:
I don't understand. If we do:

```
set_head("foo", 0x1);
set_head("foo", 0x2);
```

Can't we just run the algorithm with `old = <original value of "foo", before this tx>` and `new = 0x2`?

Reply:
I think you are right... just need to figure out how to get the original value. I could get it the first time `set_head("foo")` is called and just update the following `set_head("foo")` calls to not overwrite the old... maybe there is a way to get the underlying value from the kv?

Reply:
You can get it on first write and store it. I don't think it makes sense to expose it from kv::Write.